app/src/org/commcare/android/database/user/UserDatabaseUpgrader.java

```diff
@@ -202,6 +214,62 @@ private boolean upgradeSevenEight(SQLiteDatabase db, int oldVersion, int newVers
         }
     }
 
+    /**
+     * Adding an appId field to FormRecords
+     */
+    private boolean upgradeEightNine(SQLiteDatabase db) {
+        db.beginTransaction();
+        try {
+            if (multipleInstalledAppRecords()) {
+                // Cannot migrate FormRecords once this device has already started installing multiple
+                // applications, because there is no way to know which of those apps the existing
+                // FormRecords belong to
+                throw new MigrationException(true);
```

Comment:
What is the end result of this exception being raised? Does CC crash or is the user told to uninstall all their apps and try again? I think an alright failure mode (for power users) would be to fail silently during the migration and warn the user when they login that they need to re-install everything to fix a bug.